finetuner-workflow/finetuner/finetuner.py

```diff
@@ -46,61 +62,286 @@
+    default=False,
+    action="store_true",
+    help="Force training in fp16.",
+)
+parser.add_argument(
+    "--no_shuffle",
+    dest="no_shuffle",
+    default=False,
+    action="store_true",
+    help="Disable shuffling contexts",
+)
+parser.add_argument(
+    "--prompt_file", type=str, help="Prompt file to use for checkpoint sampling"
+)
+parser.add_argument(
+    "--prompt_every", type=int, default=0, help="Prompt every N steps"
+)
+parser.add_argument(
+    "--prompt_tokens",
+    type=int,
+    default=200,
+    help="Number of tokens to sample from prompt",
+)
+parser.add_argument(
+    "--prompt_samples",
+    type=int,
+    help="Number of samples to generate",
+    default=5,
+)
+parser.add_argument(
+    "--top_k", type=int, help="Top K to use for prompt sampling", default=50
+)
+parser.add_argument(
+    "--top_p", type=float, help="Top P to use for prompt sampling", default=0.95
+)
+parser.add_argument(
+    "--temperature",
+    type=float,
+    help="Temperature to use for prompt sampling",
+    default=1.0,
+)
+parser.add_argument(
+    "--repetition_penalty",
+    type=float,
+    help="Repetition penalty to use for prompt sampling",
+    default=1.1,
+)
+parser.add_argument(
+    "--local_rank",
+    type=int,
+    help="For distributed training: local_rank",
+    default=-1,
+)
+parser.add_argument(
+    "--fp16-full-eval",
+    dest="fp16_full_eval",
+    default=False,
+    action="store_true",
+)
+parser.add_argument(
+    "--log-level",
+    type=str,
+    help="Log level to use",
+    default="INFO",
+    choices=["DEBUG", "INFO", "WARNING", "ERROR", "CRITICAL"],
+)
 args = parser.parse_args()
 
-def no_init(loading_code: Callable[[], PreTrainedModel]) -> PreTrainedModel:
+logger = logging.getLogger(__name__)
+logger.setLevel(args.log_level.upper())
+fh = logging.StreamHandler()
+if args.local_rank != -1:
+    fh_formatter = logging.Formatter(
+        f"%(asctime)s %(levelname)s %(filename)s(%(process)d) - RANK {args.local_rank}"
+        " - %(message)s"
+    )
+else:
+    fh_formatter = logging.Formatter("%(asctime)s %(levelname)s - %(message)s")
+fh.setFormatter(fh_formatter)
+logger.addHandler(fh)
+
+
+def is_main_process() -> bool:
+    return args.local_rank in [-1, 0]
+
+
+# To be used in cases where using if statements are ugly.
+def main_process_print(*args, **kwargs):
+    if is_main_process():
+        logger.info(*args, **kwargs)
+
+
+# Where we write our training checkpoints and final model.
+output_dir = os.path.abspath(
+    os.path.join(args.output_path, "results-" + args.run_name)
+)
+
+# Properly type-cast the param (str to bool)
+FALSE = [
+    "false",
+    "f",
+    "0",
+]
+args.no_resume = args.no_resume.lower() not in FALSE
+
+# Discover if we have any checkpoints to resume from.
+if not args.no_resume:
+    try:
+        output_dir_list = os.listdir(output_dir)
+        lastCheckpoint = sorted(
+            output_dir_list, key=lambda x: int(x.split("-")[1])
+        )[-1]
+    except:
```

Comment:
A bare `except:` swallows unusual things like `SystemExit` and `KeyboardInterrupt`, so `except Exception:` ought to be used in parts like this if there is no more specific exception type known.
```suggestion
    except Exception:
```